disnake/ui/action_row.py

```diff
@@ -24,86 +24,144 @@
+    overload,
+)
+
+from ..components import (
+    ActionRow as ActionRowComponent,
+    Button as ButtonComponent,
+    SelectMenu as SelectComponent,
+    SelectOption,
+)
 from ..enums import ButtonStyle, ComponentType, TextInputStyle
 from ..utils import MISSING
 from .button import Button
 from .item import WrappedComponent
 from .select import Select
 from .text_input import TextInput
 
-__all__ = ("ActionRow",)
-
 if TYPE_CHECKING:
     from ..emoji import Emoji
+    from ..message import Message
     from ..partial_emoji import PartialEmoji
     from ..types.components import ActionRow as ActionRowPayload
 
-    ActionRowT = TypeVar("ActionRowT", bound="ActionRow")
-    Components = Union[
-        ActionRowT,
-        WrappedComponent,
-        Sequence[Union[ActionRowT, WrappedComponent, Sequence[WrappedComponent]]],
-    ]
+__all__ = ("ActionRow",)
+
 
+MessageUIComponent = Union[Button[Any], Select[Any]]
+ModalUIComponent = TextInput
+UIComponentT = TypeVar("UIComponentT", bound=WrappedComponent)
+UIComponentT_strict = TypeVar("UIComponentT_strict", MessageUIComponent, ModalUIComponent)
 
-class ActionRow:
+Components = Union[
+    "ActionRow[UIComponentT]",
+    UIComponentT,
+    Sequence[Union["ActionRow[UIComponentT]", UIComponentT, Sequence[UIComponentT]]],
+]
+
+
+class ActionRow(Generic[UIComponentT]):
     """Represents a UI action row. Useful for lower level component manipulation.
 
+    .. container:: operations
+
+        .. describe:: x[i]
+
+            Returns the component at position ``i``. This takes component width into account.
+
+    To handle interactions created by components sent in action rows or entirely independently,
+    event listeners must be used. For buttons and selects, the related events are
+    :func:`disnake.on_button_click` and :func:`disnake.on_dropdown`, respectively. Alternatively,
+    :func:`disnake.on_message_interaction` can be used for either. For modals, the related event is
+    :func:`disnake.on_modal_submit`.
+
     .. versionadded:: 2.4
 
+    .. versionchanged:: 2.6
+        Requires and provides stricter typing for contained components.
+
     Parameters
     ----------
     *components: :class:`WrappedComponent`
         The components of this action row.
+
+        .. versionchanged:: 2.6
+            Components can now be either valid in the context of a message, or in the
+            context of a modal. Combining components from both contexts is not supported.
     """
 
-    def __init__(self, *components: WrappedComponent):
-        self.width: int = 0
-        raw_components: List[NestedComponent] = []
-        # Validate the components
-        for component in components:
-            if not isinstance(component, WrappedComponent):
-                raise ValueError("ActionRow must contain only WrappedComponent instances")
+    type: ClassVar[Literal[ComponentType.action_row]] = ComponentType.action_row
 
-            self.width += component.width
+    @overload
```

Comment:
There should be an empty overload for actionrows as well, as this currently typechecks just fine:
```py
a = disnake.ui.ActionRow()
a.append_item(42)
```

Reply:
Done, alongside allowing `Select`s inside the `ModalUIComponent` type-alias, as per the new discord feature. As a result, we now get:
```py
ActionRow()                                               # ActionRow[WrappedComponent]
ActionRow(Button())                                       # ActionRow[MessageUIComponent]
ActionRow(Select())                                       # ActionRow[MessageUIComponent]
ActionRow(Button(), Select())                             # ActionRow[MessageUIComponent]
ActionRow(TextInput(label="a", custom_id="b"))            # ActionRow[ModalUIComponent]
ActionRow(Select(), TextInput(label="a", custom_id="b"))  # ActionRow[ModalUIComponent]
```
Any other combinations or types do not type-check properly.